### PR TITLE
Core Lightning: CLNRest backend

### PR DESCRIPTION
**Pull request by niteshbalusu11**
# Description

Relates to issue: #1637

Depracates c-lightning-rest and adds a new backend using core-lightning team maintained CLN Rest.
Important Note: Minimum core-lightning version to use this is `v24.02` due to breaking changes from cln team.
This PR should cover all features which the old backend was covering.
I have not tested this on iOS coz I wasn't able to get xcode to build zeus.

This pull request is categorized as a:

- [X] New feature
- [ ] Bug fix
- [ ] Code refactor
- [ ] Configuration change
- [ ] Locales update
- [ ] Quality assurance
- [ ] Other

## Checklist
- [ ] I’ve run `yarn run tsc` and made sure my code compiles correctly
- [X] I’ve run `yarn run lint` and made sure my code didn’t contain any problematic patterns
- [X] I’ve run `yarn run prettier` and made sure my code is formatted correctly
- [X] I’ve run `yarn run test` and made sure all of the tests pass

## Testing

If you modified or added a utility file, did you add new unit tests?

- [ ] No, I’m a fool
- [ ] Yes
- [ ] N/A

I have tested this PR on the following platforms (please specify OS version and phone model/VM):

- [X] Android
- [ ] iOS

### CODE REVIEW

File: backends/CLightningREST.ts
Comment (kaloudis) [inline backends/CLightningREST.ts:1]: thanks for the fixes here!

File: backends/CoreLightningRestApi.ts
Comment (kaloudis) [inline backends/CoreLightningRestApi.ts:1]: nit: let's rename this file and class `CLNRest`

Comment (kaloudis) [reply backends/CoreLightningRestApi.ts:1]: We can leave the name of the request handler file as is, as we might use it with other CLN interfaces

File: stores/SettingsStore.ts [hunk 1]
```diff
@@ -205,12 +206,26 @@ export const INTERFACE_KEYS = [
     { key: 'Embedded LND', value: 'embedded-lnd' },
     { key: 'LND (REST)', value: 'lnd' },
     { key: 'LND (Lightning Node Connect)', value: 'lightning-node-connect' },
-    { key: 'Core Lightning (c-lightning-REST)', value: 'c-lightning-REST' },
+    { key: 'Core Lightning REST API', value: 'core-lightning-rest-api' },
```
Comment (kaloudis) [inline stores/SettingsStore.ts:209]: nit: let's call the key `Core Lightning (CLNRest)` and the value `cln-rest`

**Comment by niteshbalusu11** [comment]
Fixed and rebased

File: backends/CLNRest.ts [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+import Base64Utils from '../utils/Base64Utils';
+import { Hash as sha256Hash } from 'fast-sha256';
+import BigNumber from 'bignumber.js';
+import {
+    getBalance,
+    getOffchainBalance,
+    listPeers
+} from './CoreLightningRequestHandler';
+
+export default class CLNRest extends LND {
```
Comment (vincenzopalazzo) [inline backends/CLNRest.ts:15]:
```suggestion
export default class CLNRest  {
```

There is no reason to extend LND, but we should re-export the method `getUrl` somewhere-else

**Review by vincenzopalazzo — CHANGES_REQUESTED** [review]
Ok looks like that you arrive before me, I left some comments because in this way I can build on top of you

File: backends/CLNRest.ts [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    supportsOnchainReceiving = () => true;
+    supportsLightningSends = () => true;
+    supportsKeysend = () => true;
+    supportsChannelManagement = () => true;
+    supportsPendingChannels = () => false;
+    supportsMPP = () => false;
+    supportsAMP = () => false;
+    supportsCoinControl = () => this.supports('v0.8.2', undefined, 'v0.4.0');
+    supportsChannelCoinControl = () =>
+        this.supports('v0.8.2', undefined, 'v0.4.0');
```
[the remark this answers is not in the document]
Comment (niteshbalusu11) [reply backends/CLNRest.ts:254]: These are old functions that existed in the previous c-lightning-rest and i copied them over. Probably will leave them as is.

File: views/Settings/NodeConfiguration.tsx [hunk 4]
```diff
@@ -1347,6 +1357,75 @@ export default class NodeConfiguration extends React.Component<
                                 )}
                             </>
                         )}
+                        {implementation === 'cln-rest' && (
```
[the remark this answers is not in the document]
Comment (niteshbalusu11) [reply views/Settings/NodeConfiguration.tsx:1360]: Because cln-rest uses runes and the old one uses a macaroon. We can just delete the old when Zeus wants to depracate it completely. Right now both are supported.

File: backends/CLNRest.ts (hunk 2 above)
Comment (niteshbalusu11) [reply backends/CLNRest.ts:15]: Some views are breaking when we not extend from lnd, also all the actual execution of the fetch calls happen in the LND class so its not just getUrl. Will leave it as is.

File: backends/CLNRest.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaloudis) [reply backends/CLNRest.ts:251]: I think you may have to delete the third parameter as that is an API version check

File: views/Settings/NodeConfiguration.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply views/Settings/NodeConfiguration.tsx:1360]: >Because cln-rest uses runes and the old one uses a macaroon.  

are the same things a the concept level, you can just replace the string with `implementation === 'cln-rest' ? "rune" : "macaroon"` no?

File: backends/CLNRest.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (vincenzopalazzo) [reply backends/CLNRest.ts:251]: can you clarify at least what `v0.8.2` is?

[the remark this answers is not in the document]
Comment (niteshbalusu11) [reply backends/CLNRest.ts:251]: Maybe we just mark it as true since its being supported from a very old version and the new cln-rest we are only supporting from v24.0.2 anyway?

File: backends/CLNRest.ts (hunk 2 above)
Comment (vincenzopalazzo) [reply backends/CLNRest.ts:15]: >also all the actual execution of the fetch calls happen in the LND class so its not just getUrl. 

What do you mean? This approach will leave many calls unused, and you'll likely need to rewrite a lot of code as well when you support commando (or you are not supporting it?).

TL;DR: I disagree with using CLN code that relies on LND. Even though the current API design is different, the previous design wasn't bad because c-lightning-REST emulates LND. However, clnrest/commando does not.

File: views/Settings/NodeConfiguration.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (niteshbalusu11) [reply views/Settings/NodeConfiguration.tsx:1360]: I just find them cleaner to be separate because old c-lightning-rest is sharing with now anyway. Whatever Evan wants.

File: backends/CLNRest.ts (hunk 2 above)
Comment (niteshbalusu11) [reply backends/CLNRest.ts:15]: We are not supporting Commando in this PR. 
The only thing the CLN class relies on LND is the actual "fetch" function to call the APIs. everything else is separate. I tried removing the LND class extension but that was breaking a couple of UI components. I didn't find the time to investigate that.

Comment (niteshbalusu11) [reply backends/CLNRest.ts]: Nvm, i figured out the issue. CLN no longer extends LND now.

**Comment by sha-265** [comment]
you are too fast 😵‍💫 

anyway, this is probably a good opportunity to fix `getTransactions()` which uses wrong API method ('listtransactions`), as i did in my version:
https://github.com/sha-265/zeus/blob/c40d9c1ba20c504f5e3355cf33d9e8b4409fca09/backends/CLNRest.ts#L33

**Comment by niteshbalusu11** [comment]
> anyway, this is probably a good opportunity to fix `getTransactions()` that uses wrong api method (`listfunds` which return the wallet utxos, and not transactions history), as i did in my version: https://github.com/sha-265/zeus/blob/c40d9c1ba20c504f5e3355cf33d9e8b4409fca09/backends/CLNRest.ts#L33

Thank you for catching that. I think your code is missing tag == `channel_open` and tag == `channel_close` which idk if cln also records has deposit and withdrawal. also it records chain_fee separately. 
Have to look into that.

**Comment by newtonick** [comment]
I get this error when trying send onchain in signet via CLNRest (via this PR)

"unknown parameter: address, this may be caused by
a failure to autodetect key=value-style parameters.
Please try using the -k flag and explicit key=value
pairs of parameters."

image
image

File: backends/CLNRest.ts [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+        });
+    getLightningBalance = () =>
+        this.postRequest('/v1/listfunds').then((res) => {
+            return getOffchainBalance(res);
+        });
+    sendCoins = (data: TransactionRequest) => {
+        let request: any;
+        if (data.utxos) {
+            request = {
+                address: data.addr,
```
Comment (newtonick) [inline backends/CLNRest.ts:200]: should address be destination? https://docs.corelightning.org/reference/lightning-withdraw

Comment (niteshbalusu11) [reply backends/CLNRest.ts]: Thank you for catching that. fixed.

File: backends/CLNRest.ts [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+                feeRate: `${Number(data.sat_per_vbyte) * 1000}perkb`,
```
Comment (newtonick) [inline backends/CLNRest.ts:201]: `feeRate` should also be `feerate` and `satoshis` should also be `satoshi`. Sorry I didn't catch that when I saw address was wrong.

Comment (niteshbalusu11) [reply backends/CLNRest.ts]: No, not your fault. I missed them. I fixed them now.

**Comment by sha-265** [comment]
> > anyway, this is probably a good opportunity to fix `getTransactions()` that uses wrong api method (`listfunds` which return the wallet utxos, and not transactions history), as i did in my version: https://github.com/sha-265/zeus/blob/c40d9c1ba20c504f5e3355cf33d9e8b4409fca09/backends/CLNRest.ts#L33
> 
> Thank you for catching that. I think your code is missing tag == `channel_open` and tag == `channel_close` which idk if cln also records has deposit and withdrawal. also it records chain_fee separately. Have to look into that.

i don't see this tags (at least under `wallet` account), but there is a `journal_entry` tag that should be filtered out maybe.

**Comment by sha-265** [comment]
> > i don't see this tags (at least under `wallet` account), but there is a `journal_entry` tag that should be filtered out maybe.
> 
> Yeah, that's the problem, it doesn't put it under wallet. Have to figure out where it does. In the examples in the docs, it puts it under some long string account. Have to see what that is.

it's channel id according to the cln documentation:
> account (string): The account name. If the account is a channel, the channel_id.

i'm not sure this events are relevant if you want only the on-chain transactions.

**Comment by niteshbalusu11** [comment]
> i'm not sure these events are relevant if you want only the on-chain transactions.

I think they are relevant because channel opens and closes are chain txs. I pushed a commit to include those. Please test if you can. I just pushed a commit to make the txs more readable so you know what the tx is for with a new notes field.

**Comment by sha-265** [comment]
> > i'm not sure these events are relevant if you want only the on-chain transactions.
> 
> I think they are relevant because channel opens and closes are chain txs. I pushed a commit to include those. Please test if you can. I just pushed a commit to make the txs more readable so you know what the tx is for with a new notes field.

i don't know if there is no duplicate events in `wallet` account for channel opening (at least). 

but i think in terms of performance query with out `account` parameters would be much more heavy as it returns all the non on-chain events, that could be a lot of events, so it should be considered also.

**Comment by niteshbalusu11** [comment]
> but i think in terms of performance query without account parameter, would be much more heavy as it returns all the non on-chain events

I agree. CLNs APIs don't seem to allow pagination as well which is annoying. Maybe someone can test with their mainnet node which is old and see how it performs. 
They should ideally have a simple RPC for chain txs but I'm not gonna rant about their RPC build choices here.

**Comment by sha-265** [comment]
> > but i think in terms of performance query without account parameter, would be much more heavy as it returns all the non on-chain events
> 
> I agree. CLNs APIs don't seem to allow pagination as well which is annoying. Maybe someone can test with their mainnet node which is old and see how it performs. They should ideally have a simple RPC for chain txs but I'm not gonna rant about their RPC build choices here.

yep, not the best api i've come across 🤷‍♂️

other solution could be somehow to make only one query for `getTransactions()` and `getInvoices()` instead of two separated calls for the same data.

**Comment by newtonick** [comment]
From what testing I have been able to do, I believe activity view needs some love. It might be helpful to start with what transactions should be included in this view. I'm not sure if this is documented somewhere (like a previous zeus GitHub issue/PR/etc). What I expect to be listed in the activity view is:

- lightning Invoices (paid, unpaid, expired)
- lightning payment (successful payments only) w/ fees
- onchain withdraw/payment (in mempool and onchain) w/ fees
- onchain receives (in mempool and onchain)

I don't expect to see channel open / close transactions but I also don't mind if they are in this view if it's communicated that they are part of an open/close channel transaction.

Right now I see multiple rows in the activity view for a single on chain payments. I have a hard time dissecting what information bkpr-listaccountevents is returning. Sorry this comment is not more helpful. I just wanted to provide feedback on what I have tested so far with this PR. Thank you for this work.

**Comment by niteshbalusu11** [comment]
> What I expect to be listed in the activity view is:

lightning Invoices (paid, unpaid, expired)
lightning payment (successful payments only) w/ fees
onchain withdraw/payment (in mempool and onchain) w/ fees
onchain receives (in mempool and onchain)

This sounds good to me except expired and unpaid invoices, that could be a bit too much specially expired. 
I will leave it to @kaloudis once he tests and gives feedback on what he thinks.

**Comment by newtonick** [comment]
Here is an example of 40,000 sat onchain spend (https://mutinynet.com/tx/99a918361919956cc76801b6014e233d834466dbb25b4f031825e1eb60b972e8) that shows 3 rows in the activity view for a single spend transaction. The 40,000 sat row appears first as soon as I complete the spend (while the tx is in the mempool). This shows as a credit but is actually a debit. Then once the transaction is confirmed into 1+ blocks, 2 more rows appear. A debit of the utxo I used to make the spend and then a credit of the change back to my wallet.

image

**Comment by niteshbalusu11** [comment]
> Here is an example of 40,000 sat onchain spend (https://mutinynet.com/tx/99a918361919956cc76801b6014e233d834466dbb25b4f031825e1eb60b972e8) that shows 3 rows in the activity view for a single spend transaction. The 40,000 sat row appears first as soon as I complete the spend (while the tx is in the mempool). This shows as a credit but is actually a debit. Then once the transaction is confirmed into 1+ blocks, 2 more rows appear. A debit of the utxo I used to make the spend and then a credit of the change back to my wallet.
> 
> image
> 

Ok, yeah this needs to be fixed. Only the spend should show up, not the utxos spent and change received.

**Comment by sha-265** [comment]
> I don't expect to see channel open / close transactions but I also don't mind if they are in this view if it's communicated that they are part of an open/close channel transaction.

i agree, and it's also solve the potential performance issue, as you can request only `wallet` account events, and not all the events for all the accounts in this node.

**Comment by niteshbalusu11** [comment]
ok @newtonick can you test now? I think it should be fixed.

@sha-265 unfortunately i think we can't get around the performance issue without compromising accuracy of the info because CLN displays change outputs as deposits and its really difficult to figure out if its a real deposit or a change output from like channel opens without checking channel opens from events. I asked in the CLN group if there is an easy way to check if an address is ours in CLN so that we can easily filter out change, let's see.

**Comment by sha-265** [comment]
i don't understand how tons of irrelevant events solving this, but at least you should get the same data only once and not three different times.

**Comment by niteshbalusu11** [comment]
> i don't understand how tons of irrelevant events solving this, but at least you should get the same data only once and not three different times.

I don't understand what you mean? I am getting only once.

**Comment by sha-265** [comment]
look here:
https://github.com/ZeusLN/zeus/blob/e89187c68c6de95487ce94267849e1c1b416e9c5/stores/ActivityStore.ts#L165-L168

i think if you are using `bkpr-listaccountevents` this three calls are overlapping.

**Comment by niteshbalusu11** [comment]
> look here:
> https://github.com/ZeusLN/zeus/blob/e89187c68c6de95487ce94267849e1c1b416e9c5/stores/ActivityStore.ts#L165-L168
> 
> i think if you are using `bkpr-listaccountevents` this three calls are overlapping.

Ah, that's old code. I never touched that. Will look into it.

**Comment by niteshbalusu11** [comment]
Ok I was right. It is what it is with CLN.

**Comment by niteshbalusu11** [comment]
https://github.com/ElementsProject/lightning/issues/7397

**Comment by Sjors** [comment]
I finally got the clnrest plugin to work (Python environments are a pain...). Happy to test all this when there's a testflight build.

File: backends/CLightningREST.ts [hunk 7]
```diff
@@ -10,8 +10,7 @@ import BigNumber from 'bignumber.js';
 export default class CLightningREST extends LND {
     getHeaders = (macaroonHex: string): any => {
         return {
-            macaroon: macaroonHex,
-            encodingtype: 'hex'
```
Comment (kaloudis) [inline backends/CLightningREST.ts:14]: can we revert this change? this will break connections for people with existing configurations

**Review by kaloudis — CHANGES_REQUESTED** [review]

**Review by kaloudis — APPROVED** [review]
tACK

All things looking good. Congrats, @niteshbalusu11 you will be receiving the bounty.

We are expecting you to help address any issues that may arise during alpha and beta testing - not expecting anything major as we've already done some extensive testing

**Comment by Sjors** [comment]
Just tested the https://github.com/ZeusLN/zeus/releases/tag/v0.9.0-beta1 release on iOs. It manages to connect!

I can seen my utxos on the coins screen. I can see my channels.

I can create an invoice to receive. The screen doesn't update when the invoice is paid, but afaik it didn't do that with c-lightning-rest either.

The Activity tab still crashes as it did with c-lightning-rest.

The combination of the above two things means I can't see if a payment to me succeeded.

I can pay a regular bolt11 invoice.

**Comment by niteshbalusu11** [comment]
> The Activity tab still crashes as it did with c-lightning-rest.

@Sjors is your node really old with a lot of data?

**Comment by Sjors** [comment]
@niteshbalusu11 the node is very old, but it does run the latest v24.05. It has tens of 10K+ transactions, mostly keysends from podcasting 2.0 listeners.

I guess it's tripping over missing field in older transactions?

**Comment by niteshbalusu11** [comment]
> I guess it's tripping over missing field in older transactions?

It could also be that the response from the 10k+ txns could be very large because the RPC call right now has no way limit the records returned. Working on figuring that part out.

**Comment by niteshbalusu11** [comment]
This *might* fix some issues.

https://github.com/ZeusLN/zeus/pull/2290

**Comment by sha-265** [comment]
> Important Note: Minimum core-lightning version to use this is `v24.02` due to breaking changes from cln team.

I don't know how, but it's working for me with older version

**Comment by Sjors** [comment]
> The screen doesn't update when the invoice is paid, but afaik it didn't do that with c-lightning-rest either.

More recently this usually does work.

> The Activity tab still crashes as it did with c-lightning-rest.

This is fixed since.
